File: Habitica/src/main/java/com/habitrpg/android/habitica/ui/fragments/social/InboxMessageListFragment.java
package com.habitrpg.android.habitica.ui.fragments.social;

import android.app.Activity;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.content.ContextCompat;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.Editable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.TextView;

import com.habitrpg.android.habitica.R;
import com.habitrpg.android.habitica.components.AppComponent;
import com.habitrpg.android.habitica.data.SocialRepository;
import com.habitrpg.android.habitica.data.UserRepository;
import com.habitrpg.android.habitica.events.commands.SendNewInboxMessageCommand;
import com.habitrpg.android.habitica.helpers.RxErrorHandler;
import com.habitrpg.android.habitica.models.user.User;
import com.habitrpg.android.habitica.ui.activities.FullProfileActivity;
import com.habitrpg.android.habitica.ui.adapter.social.ChatRecyclerViewAdapter;
import com.habitrpg.android.habitica.ui.fragments.BaseMainFragment;
import com.habitrpg.android.habitica.ui.helpers.SafeDefaultItemAnimator;
import com.habitrpg.android.habitica.ui.helpers.UiUtils;

import org.greenrobot.eventbus.Subscribe;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.OnTextChanged;
import io.realm.RealmResults;
import io.realm.Sort;

public class InboxMessageListFragment extends BaseMainFragment
        implements SwipeRefreshLayout.OnRefreshListener {

    @Inject
    SocialRepository socialRepository;
    @Inject
    UserRepository userRepository;

    @BindView(R.id.recyclerView)
    RecyclerView recyclerView;
    @BindView(R.id.refresh_layout)
    SwipeRefreshLayout swipeRefreshLayout;
    @BindView(R.id.emoji_button)
    ImageButton emojiButton;
    @BindView(R.id.chat_edit_text)
    EditText chatEditText;
    @BindView(R.id.send_button)
    ImageButton sendButton;
    @BindView(R.id.community_guidelines_view)
    TextView communityGuidelinesView;

    ChatRecyclerViewAdapter chatAdapter;
    String chatRoomUser;
    String replyToUserUUID;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        hideToolbar();
        disableToolbarScrolling();
        super.onCreateView(inflater, container, savedInstanceState);

        View view = inflater.inflate(R.layout.fragment_inbox_message_list, container, false);
        ButterKnife.bind(this, view);
        swipeRefreshLayout.setOnRefreshListener(this);

        LinearLayoutManager layoutManager = new LinearLayoutManager(this.getActivity());
        //layoutManager.setReverseLayout(true);
        //layoutManager.setStackFromEnd(false);
        recyclerView.setLayoutManager(layoutManager);

        chatAdapter = new ChatRecyclerViewAdapter(null, true, user, false);
        chatAdapter.setSendingUser(this.user);
        recyclerView.setAdapter(chatAdapter);
        recyclerView.setItemAnimator(new SafeDefaultItemAnimator());
        compositeSubscription.add(chatAdapter.getUserLabelClickEvents().subscribe(userId -> FullProfileActivity.open(getContext(), userId), RxErrorHandler.handleEmptyError()));

        loadMessages();

        communityGuidelinesView.setVisibility(View.GONE);

        return view;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        view.invalidate();
        view.forceLayout();
    }

    private void loadMessages() {
        if (user != null && user.isManaged()) {
            userRepository.getInboxMessages(replyToUserUUID)
                    .first()
                    .subscribe(chatMessages -> this.chatAdapter.updateData(chatMessages), RxErrorHandler.handleEmptyError());
        }
    }

    @Override
    public void onDestroyView() {
        showToolbar();
        enableToolbarScrolling();
        super.onDestroyView();
    }

    @Override
    public void onDestroy() {
        socialRepository.close();
        userRepository.close();
        super.onDestroy();
    }

    @Override
    public void injectFragment(AppComponent component) {
        component.inject(this);
    }

    private void refreshUserInbox () {
        this.swipeRefreshLayout.setRefreshing(true);
        this.userRepository.retrieveUser(true)
                .subscribe(this::onUserReceived, RxErrorHandler.handleEmptyError());
    }

    @Override
    public void onRefresh() {
        this.refreshUserInbox();
    }

    public void onUserReceived(User user) {
        this.user = user;
        if (swipeRefreshLayout != null) {
            swipeRefreshLayout.setRefreshing(false);
        }
    }

    @Subscribe
    public void onEvent(SendNewInboxMessageCommand cmd) {
        socialRepository.postPrivateMessage(cmd.userToSendTo, cmd.message)
                .subscribe(postChatMessageResult -> this.refreshUserInbox(), RxErrorHandler.handleEmptyError());
        UiUtils.dismissKeyboard(getActivity());
    }

    public void setReceivingUser(String chatRoomUser, String replyToUserUUID) {
        this.chatRoomUser = chatRoomUser;
        this.replyToUserUUID = replyToUserUUID;
    }

    @OnTextChanged(R.id.chat_edit_text)
    public void onChatMessageTextChanged() {
        Editable chatText = chatEditText.getText();
        setSendButtonEnabled(chatText.length() > 0);
    }

    private void setSendButtonEnabled(boolean enabled) {
        int tintColor;
        if (enabled) {
            tintColor = ContextCompat.getColor(getContext(), R.color.brand_400);
        } else {
            tintColor = ContextCompat.getColor(getContext(), R.color.md_grey_400);
        }
        sendButton.setEnabled(enabled);
        sendButton.setColorFilter(tintColor);
    }

    @OnClick(R.id.send_button)
    public void sendChatMessage() {
        String chatText = chatEditText.getText().toString();
        if (chatText.length() > 0) {
            chatEditText.setText(null);
            socialRepository.postPrivateMessage(replyToUserUUID, chatText).subscribe(postChatMessageResult -> {
                if (recyclerView != null) {
                    recyclerView.scrollToPosition(0);
                }
            }, RxErrorHandler.handleEmptyError());
        }
    }

    @OnClick(R.id.emoji_button)
    public void openEmojiView() {

    }
}

File: Habitica/src/main/java/com/habitrpg/android/habitica/ui/fragments/social/InboxMessageListFragment.kt
package com.habitrpg.android.habitica.ui.fragments.social

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.os.Build
import android.os.Bundle
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.fragment.app.viewModels
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.paging.LoadState
import androidx.paging.PagingData
import androidx.recyclerview.widget.RecyclerView
import com.habitrpg.android.habitica.R
import com.habitrpg.android.habitica.data.SocialRepository
import com.habitrpg.android.habitica.databinding.FragmentInboxMessageListBinding
import com.habitrpg.android.habitica.extensions.addOkButton
import com.habitrpg.android.habitica.extensions.applyScrollContentWindowInsets
import com.habitrpg.android.habitica.helpers.AppConfigManager
import com.habitrpg.android.habitica.models.social.ChatMessage
import com.habitrpg.android.habitica.ui.activities.FullProfileActivity
import com.habitrpg.android.habitica.ui.activities.MainActivity
import com.habitrpg.android.habitica.ui.adapter.social.InboxAdapter
import com.habitrpg.android.habitica.ui.fragments.BaseMainFragment
import com.habitrpg.android.habitica.ui.fragments.ReportBottomSheetFragment
import com.habitrpg.android.habitica.ui.helpers.SafeDefaultItemAnimator
import com.habitrpg.android.habitica.ui.viewmodels.InboxViewModel
import com.habitrpg.android.habitica.ui.views.HabiticaSnackbar
import com.habitrpg.android.habitica.ui.views.HabiticaSnackbar.Companion.showSnackbar
import com.habitrpg.android.habitica.ui.views.dialogs.HabiticaAlertDialog
import com.habitrpg.common.habitica.helpers.ExceptionHandler
import com.habitrpg.common.habitica.helpers.RecyclerViewState
import com.habitrpg.common.habitica.helpers.launchCatching
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import javax.inject.Inject
import kotlin.time.DurationUnit
import kotlin.time.toDuration

@AndroidEntryPoint
class InboxMessageListFragment : BaseMainFragment<FragmentInboxMessageListBinding>() {
    override var binding: FragmentInboxMessageListBinding? = null

    override fun createBinding(
        inflater: LayoutInflater,
        container: ViewGroup?
    ): FragmentInboxMessageListBinding {
        return FragmentInboxMessageListBinding.inflate(inflater, container, false)
    }

    @Inject
    lateinit var socialRepository: SocialRepository

    @Inject
    lateinit var configManager: AppConfigManager

    private var chatAdapter: InboxAdapter? = null
    private var isScrolledToBottom = true
    private var isFirstRefresh = true
    private var hasPendingRefresh = false

    private val viewModel: InboxViewModel by viewModels()

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        this.hidesToolbar = true
        return super.onCreateView(inflater, container, savedInstanceState)
    }

    override fun onViewCreated(
        view: View,
        savedInstanceState: Bundle?
    ) {
        showsBackButton = true
        super.onViewCreated(view, savedInstanceState)

        val layoutManager = androidx.recyclerview.widget.LinearLayoutManager(this.activity)
        layoutManager.reverseLayout = true
        layoutManager.stackFromEnd = false
        binding?.recyclerView?.layoutManager = layoutManager
        lifecycleScope.launch(ExceptionHandler.coroutine()) {
            socialRepository.getMember(viewModel.recipientID ?: viewModel.recipientUsername ?: "")
                .collect {
                    mainActivity?.title = it?.displayName
                    chatAdapter?.replyToUser = it
                }
        }
        chatAdapter = InboxAdapter(viewModel.user.value)

        binding?.recyclerView?.addOnScrollListener(
            object : RecyclerView.OnScrollListener() {
                override fun onScrolled(
                    recyclerView: RecyclerView,
                    dx: Int,
                    dy: Int
                ) {
                    super.onScrolled(recyclerView, dx, dy)
                    val wasScrolledToBottom = isScrolledToBottom
                    isScrolledToBottom = layoutManager.findFirstVisibleItemPosition() <= 3

                    if (isScrolledToBottom && !wasScrolledToBottom && hasPendingRefresh) {
                        hasPendingRefresh = false
                        hideNewMessageIndicator()
                        viewModel.invalidateDataSource()
                    }

                    if (isScrolledToBottom) {
                        hideNewMessageIndicator()
                    }
                }
            }
        )

        chatAdapter?.registerAdapterDataObserver(
            object : RecyclerView.AdapterDataObserver() {
                override fun onItemRangeInserted(
                    positionStart: Int,
                    itemCount: Int
                ) {
                    if (positionStart == 0 && (isScrolledToBottom || isFirstRefresh)) {
                        binding?.recyclerView?.scrollToPosition(0)
                    }
                }
            }
        )
        binding?.recyclerView?.adapter = chatAdapter
        binding?.recyclerView?.itemAnimator = SafeDefaultItemAnimator()
        chatAdapter?.let { adapter ->
            adapter.onOpenProfile = {
                FullProfileActivity.open(it)
            }
            adapter.onDeleteMessage = { showDeleteConfirmationDialog(it) }
            adapter.onFlagMessage = { showFlagMessageBottomSheet(it) }
            adapter.onCopyMessage = { copyMessageToClipboard(it) }
        }

        binding?.newMessageIndicator?.setOnClickListener {
            hasPendingRefresh = false
            hideNewMessageIndicator()
            binding?.recyclerView?.smoothScrollToPosition(0)
            viewModel.invalidateDataSource()
        }

        viewModel.messages.observe(viewLifecycleOwner) {
            markMessagesAsRead(it)
            lifecycleScope.launchCatching {
                chatAdapter?.submitData(it)
            }
        }

        chatAdapter?.addLoadStateListener { loadStates ->
            val isEmpty = loadStates.refresh is LoadState.NotLoading &&
                    loadStates.append.endOfPaginationReached &&
                    chatAdapter?.itemCount == 0

            if (isEmpty) {
                binding?.recyclerView?.state = RecyclerViewState.EMPTY
            } else {
                binding?.recyclerView?.state = RecyclerViewState.DISPLAYING_DATA
            }
        }

        binding?.chatBarView?.sendAction = { sendMessage(it) }
        binding?.chatBarView?.maxChatLength = configManager.maxChatLength()

        binding?.chatBarView?.hasAcceptedGuidelines = true

        lifecycleScope.launchCatching {
            lifecycle.repeatOnLifecycle(Lifecycle.State.RESUMED) {
                while (true) {
                    refreshConversation()
                    delay(30.toDuration(DurationUnit.SECONDS))
                }
            }
        }
        binding?.root.apply {
            ViewCompat.setOnApplyWindowInsetsListener(this!!) { _, insets ->
                val ime = insets.getInsets(WindowInsetsCompat.Type.ime()).bottom
                val nav = insets.getInsets(WindowInsetsCompat.Type.navigationBars()).bottom

                binding?.chatBarView?.translationY = -ime.toFloat()
                binding?.chatBarView?.setPadding(
                    binding?.chatBarView?.paddingLeft ?: 0,
                    binding?.chatBarView?.paddingTop ?: 0,
                    binding?.chatBarView?.paddingRight ?: 0,
                    nav
                )

                binding?.recyclerView?.translationY = -ime.toFloat()

                binding?.recyclerView?.setPadding(
                    binding?.recyclerView?.paddingLeft ?: 0,
                    binding?.recyclerView?.paddingTop ?: 0,
                    binding?.recyclerView?.paddingRight ?: 0,
                    ime + nav
                )

                insets
            }
            ViewCompat.requestApplyInsets(this)
        }
    }

    override fun onResume() {
        if (viewModel.recipientID?.isNotBlank() != true && viewModel.recipientUsername?.isNotBlank() != true) {
            parentFragmentManager.popBackStack()
        }
        binding?.root?.let { ViewCompat.requestApplyInsets(it) }
        super.onResume()
    }

    override fun onDestroy() {
        socialRepository.close()
        super.onDestroy()
    }

    override fun onCreateOptionsMenu(
        menu: Menu,
        inflater: MenuInflater
    ) {
        this.mainActivity?.menuInflater?.inflate(R.menu.inbox_chat, menu)
        super.onCreateOptionsMenu(menu, inflater)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.open_profile -> {
                openProfile()
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    private fun markMessagesAsRead(messages: PagingData<ChatMessage>) {
        // socialRepository.markSomePrivateMessagesAsRead(viewModel.user.value, messages)
    }

    private fun refreshConversation() {
        if (viewModel.memberID?.isNotBlank() != true) {
            return
        }
        lifecycleScope.launch(ExceptionHandler.coroutine()) {
            socialRepository.retrieveInboxMessages(viewModel.recipientID ?: "", 0)

            if (isScrolledToBottom || isFirstRefresh) {
                viewModel.invalidateDataSource()
                isFirstRefresh = false
            } else {
                hasPendingRefresh = true
                showNewMessageIndicator()
            }
        }
    }

    private fun sendMessage(chatText: String) {
        viewModel.memberID?.let { userID ->
            lifecycleScope.launch(
                ExceptionHandler.coroutine { error ->
                    ExceptionHandler.reportError(error)
                    binding?.let {
                        val alert = HabiticaAlertDialog(it.chatBarView.context)
                        alert.setTitle("You cannot reply to this conversation")
                        alert.setMessage("This user is unable to receive your private message")
                        alert.addOkButton()
                        alert.show()
                    }
                    binding?.chatBarView?.message = chatText
                }
            ) {
                socialRepository.postPrivateMessage(userID, chatText)
                delay(200.toDuration(DurationUnit.MILLISECONDS))
                viewModel.invalidateDataSource()
            }
        }
    }

    private fun copyMessageToClipboard(chatMessage: ChatMessage) {
        val clipMan = activity?.getSystemService(Context.CLIPBOARD_SERVICE) as? ClipboardManager
        val messageText = ClipData.newPlainText("Chat message", chatMessage.text)
        clipMan?.setPrimaryClip(messageText)
        val activity = activity as? MainActivity
        if (activity != null && Build.VERSION.SDK_INT <= Build.VERSION_CODES.S_V2) {
            showSnackbar(
                activity.snackbarContainer,
                getString(R.string.chat_message_copied),
                HabiticaSnackbar.SnackbarDisplayType.NORMAL
            )
        }
    }

    private fun showFlagMessageBottomSheet(chatMessage: ChatMessage) {
        val reportBottomSheetFragment =
            ReportBottomSheetFragment.newInstance(
                reportType = ReportBottomSheetFragment.REPORT_TYPE_MESSAGE,
                profileName = chatMessage.username ?: "",
                messageId = chatMessage.id,
                messageText = chatMessage.text ?: "",
                groupId = chatMessage.groupId ?: "",
                userIdBeingReported = chatMessage.userID ?: "",
                sourceView = this::class.simpleName ?: ""
            )

        reportBottomSheetFragment.show(childFragmentManager, ReportBottomSheetFragment.TAG)
    }

    private fun showDeleteConfirmationDialog(chatMessage: ChatMessage) {
        val context = context
        if (context != null) {
            val dialog = HabiticaAlertDialog(context)
            dialog.setTitle(R.string.confirm_delete_tag_title)
            dialog.setMessage(R.string.confirm_delete_tag_message)
            dialog.addButton(R.string.yes, true, true) { _, _ ->
                lifecycleScope.launchCatching {
                    socialRepository.deleteMessage(chatMessage)
                    viewModel.invalidateDataSource()
                }
            }
            dialog.addButton(R.string.no, false)
            dialog.show()
        }
    }

    private fun openProfile() {
        viewModel.memberID?.let { FullProfileActivity.open(it) }
    }

    private fun showNewMessageIndicator() {
        binding?.newMessageIndicator?.apply {
            if (visibility != View.VISIBLE) {
                visibility = View.VISIBLE
                alpha = 0f
                translationY = 50f
                animate()
                    .alpha(1f)
                    .translationY(0f)
                    .setDuration(200)
                    .start()
            }
        }
    }

    private fun hideNewMessageIndicator() {
        binding?.newMessageIndicator?.apply {
            if (visibility == View.VISIBLE) {
                animate()
                    .alpha(0f)
                    .translationY(50f)
                    .setDuration(200)
                    .withEndAction { visibility = View.GONE }
                    .start()
            }
        }
    }
}
